exonum/src/helpers/merkle.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Merkle tree utilities.
+
+use crypto::{self, Hash, HashStream};
+
+/// Compute Merkle root hash for a given list of hashes.
+///
+/// If `hashes` are empty then default (zero) value is returned.
+pub fn root_hash<'a, I: IntoIterator<Item=&'a Hash>>(hashes: I) -> Hash {
+    let mut current_hashes: Vec<Hash> = hashes.into_iter().cloned().collect();
```

Comment:
Is `cloned` really needed here? Hashes implement `Copy` afaik

Reply:
It seems so. `Hash` _does_ implement `Copy`, but without `.cloned()` tacked into the pipeline `collect()` does not seem to be smart enough to perform the copy itself. Or maybe this is done on purpose by libstd devs so that copying is explicit.

```
error[E0277]: the trait bound `std::vec::Vec<crypto::Hash>: std::iter::FromIterator<&'a crypto::Hash>` is not satisfied
  --> exonum/src/helpers/merkle.rs:23:60
   |
23 |     let mut current_hashes: Vec<Hash> = hashes.into_iter().collect();
   |                                                            ^^^^^^^ a collection of type `std::vec::Vec<crypto::Hash>` cannot be built from an iterator over elements of type `&'a crypto::Hash`
   |
   = help: the trait `std::iter::FromIterator<&'a crypto::Hash>` is not implemented for `std::vec::Vec<crypto::Hash>`
```

We expect and receive an iterator over references to Hashes in `BlockResponse::verify_tx_hash()` as its `self.transactions()` method returns `&[Hash]`.